### PR TITLE
feat(webrtc): implement fin ack

### LINKED ISSUE #4600: webrtc: support `FIN_ACK`
## Description

Implement webrtc `FIN_ACK` specification.

https://github.com/libp2p/specs/pull/582 

## Motivation

Closing data channels without data loss.

## Are you planning to do it yourself in a pull request?

No

### PR DESCRIPTION
**Pull request by tesol2y090**
## Description

Implements FIN_ACK support for WebRTC streams to ensure reliable closure handshaking as specified in the WebRTC spec. This adds:
- FIN_ACK flag handling in state machine
- Timeout mechanism for FIN_ACK wait
- Proper state transitions for graceful closure
- https://github.com/libp2p/rust-libp2p/issues/4600

## Notes & open questions

- Using `Instant` for deadline tracking instead of tokio timer to avoid extra dependencies
- Timeout is set to 10 seconds - should this be configurable?
- State machine ensures FIN_ACK is only sent/processed in appropriate states

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
  - Added documentation for FIN_ACK related functions
  - Updated state machine documentation
- [x] I have added tests that prove my fix is effective or that my feature works
  - Added tests for FIN_ACK state transitions
  - Added tests for timeout behavior
  - Added tests for edge cases

### CODE REVIEW

**Comment by elenaf9** [comment]
Thank you for your contribution @tesol2y090 and sorry for the delay here! Will try to give this a review in the next days.

File: misc/webrtc-utils/CHANGELOG.md [hunk 1]
```diff
@@ -1,20 +1,25 @@
...
-  See [PR 4968](https://github.com/libp2p/rust-libp2p/pull/4968).
+-   Update to latest version of `libp2p-noise`.
+    See [PR 4968](https://github.com/libp2p/rust-libp2p/pull/4968).
 
 ## 0.1.0
 
-- Initial release.
-  See [PR 4248].
+-   Initial release.
+    See [PR 4248].
```
Comment (elenaf9) [inline misc/webrtc-utils/CHANGELOG.md]: Please revert unrelated formatting changes :slightly_smiling_face:

File: misc/webrtc-utils/src/stream.rs [hunk 2]
```diff
@@ -231,19 +246,33 @@ where
...
-                        .send(GracefullyClosed {});
+                    // Check if we've received FIN_ACK or deadline has passed
+                    if self.state.fin_ack_received()
+                        || self
+                            .fin_ack_deadline
+                            .is_some_and(|deadline| Instant::now() >= deadline)
+                    {
+                        if !self.state.fin_ack_received() {
+                            tracing::warn!("FIN_ACK timeout, forcing close");
+                        }
```
Comment (elenaf9) [inline misc/webrtc-utils/src/stream.rs]: Can we do early returns instead of nested if-statements here?

File: misc/webrtc-utils/src/stream.rs [hunk 3]
```diff
@@ -231,19 +246,33 @@ where
...
                     ready!(self.io.poll_flush_unpin(cx))?;
 
-                    self.state.write_closed();
-                    let _ = self
-                        .drop_notifier
-                        .take()
-                        .expect("to not close twice")
-                        .send(GracefullyClosed {});
+                    // Check if we've received FIN_ACK or deadline has passed
+                    if self.state.fin_ack_received()
```
Comment (elenaf9) [inline misc/webrtc-utils/src/stream.rs]: Not sure about this, but don't we need to somehow continue reading here to be able to receive the `FIN_ACK`? 

With `AsyncWriteExt` for instance a future will be created that only returns after `poll_close` succeeded i.e. no concurrent call to `poll_read` happens. With the current impl, I don't think a `FIN_ACK` will ever be read in that case?

On the other hand, we'd then have to discard the read messages or buffer them. Not sure how this is handled e.g. in QUIC, or in the go implementation.

**Review by elenaf9 — COMMENTED** [review]
Sorry for the delay here @tesol2y090.

Some first comments.

File: misc/webrtc-utils/src/stream.rs [hunk 4]
```diff
@@ -53,6 +54,8 @@ const VARINT_LEN: usize = 2;
 const PROTO_OVERHEAD: usize = 5;
 /// Maximum length of data, in bytes.
 const MAX_DATA_LEN: usize = MAX_MSG_LEN - VARINT_LEN - PROTO_OVERHEAD;
+/// FIN_ACK timeout
+const FIN_ACK_TIMEOUT: Duration = Duration::from_secs(10);
```
Comment (elenaf9) [inline misc/webrtc-utils/src/stream.rs:58]: Is this number part of specs or was discussed somewhere? 
This is time that we'll have to wait each time when closing a stream if the remote doesn't support `FIN_ACK` right?

Comment (tesol2y090) [reply misc/webrtc-utils/src/stream.rs:58]: Yes, that's right. I follow the go-libp2p implementation. https://github.com/libp2p/go-libp2p/blob/master/p2p/transport/webrtc/stream.go#L45

File: misc/webrtc-utils/src/stream.rs (hunk 3 above)
Comment (elenaf9) [reply misc/webrtc-utils/src/stream.rs]: So I looked into the QUIC and TCP transports to check how they handle this in their `poll_close` function.
However, as far as I can tell neither of them wait in `poll_close` for the sent `FIN` to be acknowledged in the first place.

Unfortunately I couldn't find any prior discussion about this.

**Comment by elenaf9** [comment]
Regarding whether `poll_close` should wait for the `FIN_ACK` or not (see https://github.com/libp2p/rust-libp2p/pull/5687#discussion_r1898455099). 
As already mentioned, the current implementation won't work when there are blocking calls to `close().await`, because without a concurrent call to `poll_read` the corresponding `FIN_ACK` will never be read and we'll always timeout.

If the read side is closed anyway, we can just read from within `poll_close`, parse the flags, and discard any other messages [analogous to how `poll_write` is doing it](https://github.com/libp2p/rust-libp2p/blob/9907a94b3ac08e9dcf3568bf6bb58a100de090d7/misc/webrtc-utils/src/stream.rs#L180-L202).

If the read side isn't closed yet we have two options:
1. Don't wait for the `FIN_ACK`. But that would mean that we wouldn't have the guarantee that all data has been received.
2. Read from the stream and buffer the incoming messages up to a `BUFFER_MAX` (I think we can use the existing `read_buffer` for that?). If the max has been reached before the `FIN_ACK` we just return same like in case of a timeout.

I think we should go with nr. 2, but I wonder if there are any better options that I didn't think of (cc @dariusc93 @jxs).
What do you think @tesol2y090?
